language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/task/BackgroundTaskService.java

```diff
@@ -0,0 +1,161 @@
+import org.ballerinalang.langserver.commons.LanguageServerContext;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A class to track background services being run within a language server instance.
+ *
+ * @since 2.0.0
+ */
+public class BackgroundTaskService {
+
+    private static final LanguageServerContext.Key<BackgroundTaskService> TASK_SERVICE_KEY =
+            new LanguageServerContext.Key<>();
+
+    private final Map<TaskKey, CompletableFuture<Void>> tasks = new ConcurrentHashMap<>();
+
+    private BackgroundTaskService(LanguageServerContext context) {
+        context.put(TASK_SERVICE_KEY, this);
+    }
+
+    public static BackgroundTaskService getInstance(LanguageServerContext context) {
+        BackgroundTaskService taskService = context.get(TASK_SERVICE_KEY);
+        if (taskService == null) {
+            taskService = new BackgroundTaskService(context);
+        }
+
+        return taskService;
+    }
+
+    /**
+     * Submit a new task to be executed in the background. For a given package, only one task with the given name can
+     * be running at a given time.
+     *
+     * @param packageId Package with which a background task is associated
+     * @param task      The task to be executed
+     */
+    public void submit(PackageId packageId, Task task) {
+        TaskKey taskKey = TaskKey.from(packageId, task.name());
+
+        if (isRunning(packageId, task.name())) {
+            throw new IllegalStateException("A task is already running: " + taskKey);
+        }
+
+        task.setTaskId(taskKey.taskId());
+        CompletableFuture<Void> future = CompletableFuture.runAsync(() -> {
+            task.onStart();
+            try {
+                task.execute();
+            } catch (Exception e) {
+                throw new TaskExecutionException(e);
+            }
+        });
+        tasks.put(taskKey, future);
+
+        future.thenAccept(v -> {
+            task.onSuccess();
+        }).exceptionally(t -> {
+            if (t instanceof TaskExecutionException) {
+                task.onFail(t.getCause());
+            } else {
+                task.onFail(t);
+            }
+            return null;
+        }).thenApply(v -> {
+            tasks.remove(taskKey);
+            return null;
+        });
+    }
+
+    /**
+     * Check if a task with the given name is running for the provided package.
+     *
+     * @param packageId Package ID
+     * @param taskName  Name of the task
+     * @return true if a task with the given name is running for the provided package
+     */
+    public boolean isRunning(PackageId packageId, String taskName) {
+        TaskKey key = TaskKey.from(packageId, taskName);
+        return tasks.containsKey(key) && !tasks.get(key).isDone();
+    }
+
+    /**
+     * Cancels all running tasks and clear tasks map.
+     */
+    public void shutdown() {
```

Comment:
Will we have the requirement of cancelling an individua task? if not we can add it later when required

Reply:
We will have such a requirement in the future. Even for pull module code action, we should provide the ability to cancel if it takes longer. I have added to `todo` for that. Let's add that API when we implement the cancellation support.